src/Lecture4.hs

```diff
@@ -133,9 +138,56 @@ errors. We will simply return an optional result here.
 
 🕯 HINT: Use the 'readMaybe' function from the 'Text.Read' module.
 -}
-
+dropWhitespace :: [Char] -> [Char]
+dropWhitespace = takeWhile (not . isSpace) . dropWhile isSpace
+
+tokenize :: Char -> String -> [String]
+tokenize delim string = case string of
+    "" -> []
+    _ -> listOfTokens
+    where
+    go :: String -> String -> Char -> (String, String)
+    go token str delimiter = case str of
+        (x : xs) -> if x == delimiter
+                    then (token, xs)
+                    else go (token ++ [x]) xs delimiter
+        [] -> (token, [])
+
+    parsedResult = go [] string delim
+    parsedToken = fst parsedResult
+    stringRemainder = snd parsedResult
+    listOfTokens = parsedToken : tokenize delim stringRemainder
+
+toTradeType :: String -> Maybe TradeType
+toTradeType word = case dropWhitespace word of
+    "Buy" -> Just Buy
+    "Sell" -> Just Sell
+    _ -> Nothing
+
+isNatural :: Double -> Bool
+isNatural x = (x == fromInteger (round x)) && x >= 0
+
+toInt :: String -> Maybe Int
+toInt word = if isNatural (fromMaybe (-1) (readMaybe word)) then Just (read word) else Nothing
+
+{- This function is really awkward and ugly, but I have no idea how to make it neater :( 
+ - I can't really see how the andThen, maybePlus, Monad refactoring could be done here. 
+ - I could just extract the pattern matching to a separate function but that only moves 
+ - the ugly bits 5 lines up. -}
 parseRow :: String -> Maybe Row
-parseRow = error "TODO"
+parseRow row = parsedRow where
+    tokens = tokenize ',' row
+    parsedRow = case tokens of
+        [x, y, z] -> case x of
+            "" -> Nothing
+            _ -> case toTradeType y of
+                Nothing -> Nothing
+                _ -> case toInt z of
+                    Nothing -> Nothing
+                    _ -> Just Row {rowProduct=x, rowTradeType=fromJust (toTradeType y), rowCost=fromJust (toInt z)}
+        _ -> Nothing
```

Comment:
Your implementation with pattern matching is already quite good!

Indeed, as you mentioned, you can go one step further if you use the bind operator `>>=`:

```suggestion
parseRow row = parsedRow where
    tokens = tokenize ',' row
    parsedRow = case tokens of
        [x, y, z] -> case x of
            "" -> Nothing
            _ -> toTradeType y >>= \tradeType ->
                   toInt z >>= \cost ->
                   Just Row {rowProduct = x, rowTradeType = tradeType, rowCost = cost}
        _ -> Nothing
```

I haven't explained the following during lectures but there's a nice way to write this parsing syntax using the `do`-notation, `Monad` instance for `Maybe`, and the `MonadFail` desugaring trick 🧙🏻 

Applying this programming pattern and an existing helper function `guard`, the `parseRow` function will look like this:

```suggestion
parseRow row = do
    [x, y, z] <- Just $ tokenize ',' row
    guard $ x /= ""
    tradeType <- readMaybe y
    cost <- readMaybe z
    guard $ cost >= 0
    pure $ Row {rowProduct = x, rowTradeType = tradeType, rowCost = cost}
```